File: docs/static/scripts/engine/sigil-net-graph.js
/**
 * Browser inference for SigilNetGraph (graph-conv trunk).
 * Mirrors ai/sigil_net_graph.py:SigilNetGraph.forward.
 *
 * Reuses the matmul / linear / layerNorm helpers from SigilNetJS by
 * subclassing it. Overrides `forward` to take the graph path:
 *   1. Slice per-node features (39 × 11) out of the 450-dim raw vector.
 *   2. Run a single GraphConv (proj) and N GraphResBlocks over the
 *      static normalized adjacency matrix.
 *   3. Mean+max pool → board summary.
 *   4. Concat with spell embeddings + global features.
 *   5. Dense ResBlocks at DENSE_TRUNK_DIM.
 *   6. Heads (value, policy via dot product with turn projection).
 *
 * Adjacency is precomputed once at load time from constants.js
 * NODE_ORDER + ADJACENCY (must match Python's _build_normalized_adjacency).
 */

const NUM_NODES_GRAPH = 39;
const _LIFE_START_GRAPH = 250;        // matches ai/sigil_net_graph.py
const _SPELL_FILL_START_GRAPH = 406;  // 250 + 156

function _buildNormalizedAdjacency() {
	const idx = {};
	NODE_ORDER.forEach((n, i) => { idx[n] = i; });
	const A = new Float32Array(NUM_NODES_GRAPH * NUM_NODES_GRAPH);
	for (const n of NODE_ORDER) {
		const i = idx[n];
		A[i * NUM_NODES_GRAPH + i] = 1.0; // self-loop
		for (const nb of (ADJACENCY[n] || [])) A[i * NUM_NODES_GRAPH + idx[nb]] = 1.0;
	}
	const deg = new Float32Array(NUM_NODES_GRAPH);
	for (let i = 0; i < NUM_NODES_GRAPH; i++) {
		let s = 0;
		for (let j = 0; j < NUM_NODES_GRAPH; j++) s += A[i * NUM_NODES_GRAPH + j];
		deg[i] = s;
	}
	const dInvSqrt = new Float32Array(NUM_NODES_GRAPH);
	for (let i = 0; i < NUM_NODES_GRAPH; i++) {
		dInvSqrt[i] = deg[i] > 0 ? 1.0 / Math.sqrt(deg[i]) : 0;
	}
	const out = new Float32Array(NUM_NODES_GRAPH * NUM_NODES_GRAPH);
	for (let i = 0; i < NUM_NODES_GRAPH; i++) {
		for (let j = 0; j < NUM_NODES_GRAPH; j++) {
			out[i * NUM_NODES_GRAPH + j] =
				dInvSqrt[i] * A[i * NUM_NODES_GRAPH + j] * dInvSqrt[j];
		}
	}
	return out;
}

function _buildStaticNodeIndicators() {
	// Two channels: is_mana, is_spell_position. Must match Python's
	// _IS_MANA / _IS_SPELL_POS in ai/sigil_net_graph.py.
	const idx = {};
	NODE_ORDER.forEach((n, i) => { idx[n] = i; });
	const isMana = new Float32Array(NUM_NODES_GRAPH);
	const isSpell = new Float32Array(NUM_NODES_GRAPH);
	for (const n of ['a1', 'b1', 'c1']) isMana[idx[n]] = 1.0;
	for (let slot = 1; slot <= 9; slot++) {
		const nodes = POSITIONS[slot] || [];
		for (const n of nodes) isSpell[idx[n]] = 1.0;
	}
	return { isMana, isSpell };
}

class SigilNetGraphJS extends SigilNetJS {
	constructor(config, weights) {
		super(config, weights);
		this._A = _buildNormalizedAdjacency();
		const ind = _buildStaticNodeIndicators();
		this._isMana = ind.isMana;
		this._isSpell = ind.isSpell;
	}

	/**
	 * Sparse-aware aggregation: out[i] = sum_j A[i,j] * x[j].
	 * x: (NUM_NODES_GRAPH × dim). Returns same shape.
	 */
	_aggregate(x, dim) {
		const N = NUM_NODES_GRAPH;
		const out = new Float32Array(N * dim);
		for (let i = 0; i < N; i++) {
			for (let j = 0; j < N; j++) {
				const w = this._A[i * N + j];
				if (w === 0) continue;
				for (let c = 0; c < dim; c++) {
					out[i * dim + c] += w * x[j * dim + c];
				}
			}
		}
		return out;
	}

	_graphConv(x, inDim, outDim, prefix) {
		const N = NUM_NODES_GRAPH;
		// Aggregate along normalized adjacency.
		const agg = this._aggregate(x, inDim);
		// Linear + LayerNorm + ReLU.
		let h = this._linear(agg, N, inDim,
		                     `${prefix}.lin.weight`, `${prefix}.lin.bias`);
		h = this._layerNorm(h, N, outDim,
		                    `${prefix}.ln.weight`, `${prefix}.ln.bias`);
		return this._relu(h);
	}

	_graphResBlock(x, dim, prefix) {
		const h1 = this._graphConv(x, dim, dim, `${prefix}.g1`);
		const h2 = this._graphConv(h1, dim, dim, `${prefix}.g2`);
		const out = new Float32Array(x.length);
		for (let i = 0; i < x.length; i++) {
			const val = h2[i] + x[i];
			out[i] = val > 0 ? val : 0;
		}
		return out;
	}

	/**
	 * Per-node feature tensor (NUM_NODES × NODE_FEATURE_DIM=11).
	 * Layout from raw — must match SigilNetGraph._node_features in Python:
	 *   - stones_own, stones_enemy, stones_empty (3) — interleaved per node
	 *   - nbhd own_frac, enemy_frac (2) — interleaved per node
	 *   - life: own_escape, enemy_escape, own_crush, enemy_crush (4)
	 *   - static: is_mana, is_spell_position (2)
	 */
	_nodeFeatures(raw) {
		const N = NUM_NODES_GRAPH;
		const C = this.config.node_feature_dim;
		const out = new Float32Array(N * C);
		// Stones (raw[0..117]) — layout: own[39], enemy[39], empty[39]
		for (let i = 0; i < N; i++) {
			out[i * C + 0] = raw[i];           // own
			out[i * C + 1] = raw[N + i];       // enemy
			out[i * C + 2] = raw[2 * N + i];   // empty
		}
		// Neighborhood (raw[117..195]) — interleaved (own, enemy) per node
		for (let i = 0; i < N; i++) {
			out[i * C + 3] = raw[3 * N + 2 * i];     // own_frac
			out[i * C + 4] = raw[3 * N + 2 * i + 1]; // enemy_frac
		}
		// Life (raw[250..406]) — layout: own_escape[39], enemy_escape[39], own_crush[39], enemy_crush[39]
		for (let i = 0; i < N; i++) {
			out[i * C + 5] = raw[_LIFE_START_GRAPH + 0 * N + i];
			out[i * C + 6] = raw[_LIFE_START_GRAPH + 1 * N + i];
			out[i * C + 7] = raw[_LIFE_START_GRAPH + 2 * N + i];
			out[i * C + 8] = raw[_LIFE_START_GRAPH + 3 * N + i];
		}
		// Static node indicators
		for (let i = 0; i < N; i++) {
			out[i * C + 9] = this._isMana[i];
			out[i * C + 10] = this._isSpell[i];
		}
		return out;
	}

	/**
	 * Concat the parts of raw not absorbed into per-node features.
	 * Must match SigilNetGraph._global_features in Python.
	 */
	_globalFeatures(raw) {
		// head = raw[195..250) (after stones+nbhd, before life)
		// tail = raw[406..end) (after life: spell_fill, threat, tempo)
		const headLen = _LIFE_START_GRAPH - 3 * NUM_NODES_GRAPH - 2 * NUM_NODES_GRAPH; // 250 - 117 - 78 = 55
		const tailLen = raw.length - _SPELL_FILL_START_GRAPH; // 450 - 406 = 44
		const out = new Float32Array(headLen + tailLen);
		out.set(raw.subarray(3 * NUM_NODES_GRAPH + 2 * NUM_NODES_GRAPH, _LIFE_START_GRAPH), 0);
		out.set(raw.subarray(_SPELL_FILL_START_GRAPH), headLen);
		return out;
	}

	forward(rawFeatures, spellIds, turnFeatures, numTurns) {
		const cfg = this.config;
		const N = NUM_NODES_GRAPH;

		// Per-node features (39 × 11) and run through graph trunk.
		let h = this._nodeFeatures(rawFeatures);
		h = this._graphConv(h, cfg.node_feature_dim, cfg.graph_hidden_dim, 'node_proj');
		const H = cfg.graph_hidden_dim;
		for (let b = 0; b < cfg.num_graph_blocks; b++) {
			h = this._graphResBlock(h, H, `graph_blocks.${b}`);
		}

		// Pool: mean + max along the 39-node axis.
		const pooled = new Float32Array(2 * H);
		for (let c = 0; c < H; c++) {
			let s = 0, mx = -Infinity;
			for (let i = 0; i < N; i++) {
				const v = h[i * H + c];
				s += v;
				if (v > mx) mx = v;
			}
			pooled[c] = s / N;
			pooled[H + c] = mx;
		}

		// Spell embedding lookup.
		const embedTable = this._getW('spell_embed.weight');
		const embedDim = cfg.spell_embed_dim;
		const spellFlat = new Float32Array(cfg.num_spell_slots * embedDim);
		for (let i = 0; i < cfg.num_spell_slots; i++) {
			const id = spellIds[i];
			spellFlat.set(
				embedTable.subarray(id * embedDim, (id + 1) * embedDim),
				i * embedDim,
			);
		}

		// Globals (raw minus per-node).
		const globals_ = this._globalFeatures(rawFeatures);

		// Concat: [pooled | spell_emb | globals]
		const denseInDim = pooled.length + spellFlat.length + globals_.length;
		const xConcat = new Float32Array(denseInDim);
		xConcat.set(pooled, 0);
		xConcat.set(spellFlat, pooled.length);
		xConcat.set(globals_, pooled.length + spellFlat.length);

		// Dense trunk: dense_in -> LN -> ReLU -> N res blocks
		const D = cfg.dense_trunk_dim;
		let x = this._linear(xConcat, 1, denseInDim, 'dense_in.weight', 'dense_in.bias');
		x = this._layerNorm(x, 1, D, 'dense_in_ln.weight', 'dense_in_ln.bias');
		x = this._relu(x);
		for (let b = 0; b < cfg.num_dense_res_blocks; b++) {
			x = this._denseResBlock(x, 1, D, b);
		}

		// Value head
		let v = this._linear(x, 1, D, 'value_fc1.weight', 'value_fc1.bias');
		v = this._relu(v);
		v = this._linear(v, 1, cfg.value_hidden_dim, 'value_fc2.weight', 'value_fc2.bias');
		v = this._tanh(v);
		const value = v[0];

		// Policy head
		let policyLogits = null;
		if (turnFeatures && numTurns > 0) {
			const boardProj = this._linear(x, 1, D, 'policy_proj.weight', 'policy_proj.bias');
			const turnProj = this._linear(turnFeatures, numTurns, cfg.turn_feature_dim,
			                              'turn_proj.weight', 'turn_proj.bias');
			policyLogits = new Float32Array(numTurns);
			const pDim = cfg.policy_hidden_dim;
			for (let i = 0; i < numTurns; i++) {
				let dot = 0;
				for (let j = 0; j < pDim; j++) {
					dot += turnProj[i * pDim + j] * boardProj[j];
				}
				policyLogits[i] = dot;
			}
		}

		return { value, policyLogits };
	}

	_denseResBlock(x, rows, dim, blockIdx) {
		const prefix = `dense_blocks.${blockIdx}`;
		let out = this._linear(x, rows, dim, `${prefix}.fc1.weight`, `${prefix}.fc1.bias`);
		out = this._layerNorm(out, rows, dim, `${prefix}.ln1.weight`, `${prefix}.ln1.bias`);
		out = this._relu(out);
		out = this._linear(out, rows, dim, `${prefix}.fc2.weight`, `${prefix}.fc2.bias`);
		out = this._layerNorm(out, rows, dim, `${prefix}.ln2.weight`, `${prefix}.ln2.bias`);
		const result = new Float32Array(rows * dim);
		for (let i = 0; i < result.length; i++) {
			const val = out[i] + x[i];
			result[i] = val > 0 ? val : 0;
		}
		return result;
	}

	static async load(manifestUrl, binUrl) {
		const [manifestResp, binResp] = await Promise.all([
			fetch(manifestUrl),
			fetch(binUrl),
		]);
		const manifest = await manifestResp.json();
		const binBuffer = await binResp.arrayBuffer();
		const weights = {};
		for (const [key, info] of Object.entries(manifest.tensors)) {
			const arr = new Float32Array(binBuffer, info.offset, info.length);
			weights[key] = { data: arr, shape: info.shape };
		}
		return new SigilNetGraphJS(manifest.config, weights);
	}
}

File: ai/sigil_net_graph.py
"""SigilNet with a graph-convolutional trunk over the 39-node board.

The board is a sparse, fixed graph (39 nodes, max degree 4). A flat MLP
trunk has to learn this topology from scratch; here we wire it in
directly using a precomputed normalized adjacency matrix and a stack of
GCN-style message-passing layers. The graph-derived board summary is
then concatenated with the existing global features (spell embeddings,
tempo, threat, etc.) before passing through dense ResBlocks and the
value / policy / blunder heads.

Per-node features are extracted by slicing the 450-dim flat raw vector
that ai/features.py:board_to_tensor produces, so the data format is
unchanged. Adding a new model class (rather than mutating SigilNet)
keeps v22 loadable and gives us a clean A/B baseline.
"""

import os
import sys

import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from notation import NODE_ORDER, ADJACENCY, POSITIONS

from ai.config import (
    NUM_POSSIBLE_SPELLS, NUM_SPELL_SLOTS, SPELL_EMBED_DIM, NUM_NODES,
    RAW_FEATURE_DIM, TURN_FEATURE_DIM, POLICY_HIDDEN_DIM, VALUE_HIDDEN_DIM,
    MODELS_DIR,
)
from ai.sigil_net import BLUNDER_HIDDEN_DIM


# Slice indices into the 450-dim raw vector. Must match
# ai/features.py:board_to_tensor layout.
_STONES_START = 0
_STONES_LEN = NUM_NODES * 3                 # 117
_NBHD_START = _STONES_START + _STONES_LEN
_NBHD_LEN = NUM_NODES * 2                   # 78
_GLOBAL_BLOCK_END = _NBHD_START + _NBHD_LEN + 27 + 3 + 2 + 18 + 1 + 2 + 2  # = 250
_LIFE_START = _GLOBAL_BLOCK_END             # 250
_LIFE_LEN = NUM_NODES * 4                   # 156
_SPELL_FILL_START = _LIFE_START + _LIFE_LEN  # 406
_SPELL_FILL_LEN = 18
_THREAT_START = _SPELL_FILL_START + _SPELL_FILL_LEN  # 424
_THREAT_LEN = 18
_TEMPO_START = _THREAT_START + _THREAT_LEN  # 442
_TEMPO_LEN = 8

# Static per-node indicator features (constants — the same for every position).
_IS_MANA = np.zeros(NUM_NODES, dtype=np.float32)
_IS_SPELL_POS = np.zeros(NUM_NODES, dtype=np.float32)
_IDX = {n: i for i, n in enumerate(NODE_ORDER)}
for n in ('a1', 'b1', 'c1'):
    _IS_MANA[_IDX[n]] = 1.0
for slot in range(1, NUM_SPELL_SLOTS + 1):
    for n in POSITIONS.get(slot, []):
        _IS_SPELL_POS[_IDX[n]] = 1.0

# Channels assembled per-node: 3 stones + 2 nbhd + 4 life + 2 indicators = 11
NODE_FEATURE_DIM = 11

# Non-spatial features (everything in the raw vector except stones, nbhd, life):
# spell_charges(27) + mana(3) + counters(2) + lock(18) + diff(1) + counts(2)
# + turn(2) + fill(18) + threat(18) + tempo(8) = 99.
GLOBAL_FEATURE_DIM = (
    RAW_FEATURE_DIM - _STONES_LEN - _NBHD_LEN - _LIFE_LEN
)  # 450 - 117 - 78 - 156 = 99


def _build_normalized_adjacency():
    """Return (NUM_NODES, NUM_NODES) tensor: D^{-1/2} (A + I) D^{-1/2}."""
    A = np.zeros((NUM_NODES, NUM_NODES), dtype=np.float32)
    for n in NODE_ORDER:
        i = _IDX[n]
        A[i, i] = 1.0  # self-loop
        for nb in ADJACENCY.get(n, []):
            A[i, _IDX[nb]] = 1.0
    deg = A.sum(axis=1)
    d_inv_sqrt = np.where(deg > 0, 1.0 / np.sqrt(deg), 0.0)
    A_norm = (A * d_inv_sqrt).T * d_inv_sqrt    # D^-1/2 A D^-1/2
    return torch.tensor(A_norm, dtype=torch.float32)


class GraphConv(nn.Module):
    """Single GCN-style message-passing layer with LayerNorm + ReLU."""

    def __init__(self, in_dim, out_dim):
        super().__init__()
        self.lin = nn.Linear(in_dim, out_dim)
        self.ln = nn.LayerNorm(out_dim)

    def forward(self, x, A_norm):
        # x: (B, N, in_dim);  A_norm: (N, N)
        h = torch.einsum('ij,bjc->bic', A_norm, x)
        h = self.lin(h)
        h = self.ln(h)
        return F.relu(h)


class GraphResBlock(nn.Module):
    """Residual block: two GraphConv layers, summed with the input."""

    def __init__(self, dim):
        super().__init__()
        self.g1 = GraphConv(dim, dim)
        self.g2 = GraphConv(dim, dim)

    def forward(self, x, A_norm):
        h = self.g1(x, A_norm)
        h = self.g2(h, A_norm)
        return F.relu(h + x)


class SigilNetGraph(nn.Module):
    """SigilNet with a graph-conv trunk and the same heads as SigilNet."""

    GRAPH_HIDDEN_DIM = 128
    NUM_GRAPH_BLOCKS = 4   # one initial proj + this many residual blocks
    DENSE_TRUNK_DIM = 512
    NUM_DENSE_RES_BLOCKS = 3

    def __init__(self):
        super().__init__()

        # Static, non-trainable adjacency.
        self.register_buffer('A_norm', _build_normalized_adjacency())
        # Static node-type indicator broadcast per batch element.
        node_static = np.stack([_IS_MANA, _IS_SPELL_POS], axis=1)  # (39, 2)
        self.register_buffer(
            'node_static', torch.tensor(node_static, dtype=torch.float32))

        # Graph trunk.
        self.node_proj = GraphConv(NODE_FEATURE_DIM, self.GRAPH_HIDDEN_DIM)
        self.graph_blocks = nn.ModuleList([
            GraphResBlock(self.GRAPH_HIDDEN_DIM)
            for _ in range(self.NUM_GRAPH_BLOCKS)
        ])

        # Spell embedding (same as SigilNet).
        self.spell_embed = nn.Embedding(NUM_POSSIBLE_SPELLS, SPELL_EMBED_DIM)

        # Dense trunk consumes [pooled_graph (mean+max) | spell_emb_flat | globals].
        spell_flat_dim = NUM_SPELL_SLOTS * SPELL_EMBED_DIM            # 144
        pooled_dim = self.GRAPH_HIDDEN_DIM * 2                        # 192
        dense_in = pooled_dim + spell_flat_dim + GLOBAL_FEATURE_DIM   # 192+144+99=435
        self.dense_in = nn.Linear(dense_in, self.DENSE_TRUNK_DIM)
        self.dense_in_ln = nn.LayerNorm(self.DENSE_TRUNK_DIM)
        self.dense_blocks = nn.ModuleList([
            _DenseResBlock(self.DENSE_TRUNK_DIM)
            for _ in range(self.NUM_DENSE_RES_BLOCKS)
        ])

        # Heads (mirror SigilNet).
        self.value_fc1 = nn.Linear(self.DENSE_TRUNK_DIM, VALUE_HIDDEN_DIM)
        self.value_fc2 = nn.Linear(VALUE_HIDDEN_DIM, 1)
        self.policy_proj = nn.Linear(self.DENSE_TRUNK_DIM, POLICY_HIDDEN_DIM)
        self.turn_proj = nn.Linear(TURN_FEATURE_DIM, POLICY_HIDDEN_DIM)
        self.blunder_board_proj = nn.Linear(self.DENSE_TRUNK_DIM, BLUNDER_HIDDEN_DIM)
        self.blunder_turn_proj = nn.Linear(TURN_FEATURE_DIM, BLUNDER_HIDDEN_DIM)
        self.blunder_bias = nn.Parameter(torch.zeros(1))

    # --- Slicing helpers ---

    def _node_features(self, raw):
        """Extract (B, 39, NODE_FEATURE_DIM) per-node features from raw."""
        B = raw.size(0)
        # Stones: 3 channels per node, layout = own_all_nodes | enemy_all | empty_all.
        stones = raw[:, _STONES_START:_STONES_START + _STONES_LEN]
        stones = stones.view(B, 3, NUM_NODES).permute(0, 2, 1)        # (B, 39, 3)
        # Neighborhood frac: 2 per node, but layout is interleaved (own, enemy)
        # per node — see board_to_tensor.
        nbhd = raw[:, _NBHD_START:_NBHD_START + _NBHD_LEN]
        nbhd = nbhd.view(B, NUM_NODES, 2)                             # (B, 39, 2)
        # Life status: 4 per node, layout = own_escape_all | enemy_escape_all
        # | own_crush_all | enemy_crush_all  (each contiguous over 39 nodes).
        life = raw[:, _LIFE_START:_LIFE_START + _LIFE_LEN]
        life = life.view(B, 4, NUM_NODES).permute(0, 2, 1)            # (B, 39, 4)
        static = self.node_static.unsqueeze(0).expand(B, -1, -1)      # (B, 39, 2)
        return torch.cat([stones, nbhd, life, static], dim=-1)        # (B, 39, 11)

    def _global_features(self, raw):
        """Concat the parts of raw not absorbed into per-node features."""
        # Globals = bytes [195..250) ∪ [406..450).
        # Layout from board_to_tensor (after life-status block was inserted at 250):
        #   195..222  spell charges (27)
        #   222..225  mana (3)
        #   225..227  spell counters (2)
        #   227..245  lock (18)
        #   245..246  stone diff (1)
        #   246..248  total stones (2)
        #   248..250  turn progress (2)
        #   250..406  LIFE (extracted into per-node) ← skip
        #   406..424  spell fill (18)
        #   424..442  threat (18)
        #   442..450  tempo (8)
        head = raw[:, _NBHD_START + _NBHD_LEN:_LIFE_START]
        tail = raw[:, _SPELL_FILL_START:]
        return torch.cat([head, tail], dim=-1)

    def forward(self, raw_features, spell_ids, turn_features=None,
                turn_counts=None, return_blunder=False):
        node_feat = self._node_features(raw_features)                 # (B, 39, 11)
        h = self.node_proj(node_feat, self.A_norm)                    # (B, 39, H)
        for block in self.graph_blocks:
            h = block(h, self.A_norm)
        # Pool over the 39 nodes.
        pooled = torch.cat([h.mean(dim=1), h.amax(dim=1)], dim=-1)    # (B, 2H)

        spell_emb = self.spell_embed(spell_ids).reshape(spell_ids.size(0), -1)
        globals_ = self._global_features(raw_features)
        x = torch.cat([pooled, spell_emb, globals_], dim=-1)
        x = F.relu(self.dense_in_ln(self.dense_in(x)))
        for block in self.dense_blocks:
            x = block(x)

        v = F.relu(self.value_fc1(x))
        v = torch.tanh(self.value_fc2(v))

        policy_logits = None
        blunder_logits = None
        if turn_features is not None:
            expected = self.turn_proj.in_features
            if turn_features.size(-1) > expected:
                turn_features = turn_features[..., :expected]
            board_proj = self.policy_proj(x)
            turn_proj = self.turn_proj(turn_features)
            logits = torch.bmm(turn_proj, board_proj.unsqueeze(-1)).squeeze(-1)
            if turn_counts is not None:
                max_t = turn_features.size(1)
                mask = (
                    torch.arange(max_t, device=logits.device).unsqueeze(0)
                    >= turn_counts.unsqueeze(1)
                )
                logits = logits.masked_fill(mask, float('-inf'))
            policy_logits = logits

            if return_blunder:
                bb = self.blunder_board_proj(x.detach())
                tb = self.blunder_turn_proj(turn_features)
                bl = torch.bmm(tb, bb.unsqueeze(-1)).squeeze(-1) + self.blunder_bias
                if turn_counts is not None:
                    bl = bl.masked_fill(mask, 0.0)
                blunder_logits = bl

        if return_blunder:
            return v, policy_logits, blunder_logits
        return v, policy_logits

    def evaluate_with_policy(self, raw_features, spell_ids, turn_features,
                             blunder_lambda=0.0):
        raw_features = raw_features.unsqueeze(0)
        spell_ids = spell_ids.unsqueeze(0)
        turn_features = turn_features.unsqueeze(0)
        turn_counts = torch.tensor([turn_features.size(1)], dtype=torch.long)
        return_blunder = blunder_lambda > 0
        with torch.no_grad():
            out = self.forward(raw_features, spell_ids, turn_features,
                               turn_counts, return_blunder=return_blunder)
        if return_blunder:
            v, logits, b = out
            adj = logits - blunder_lambda * torch.sigmoid(b)
        else:
            v, logits = out
            adj = logits
        return v.item(), F.softmax(adj.squeeze(0), dim=0).cpu().numpy()

    def save(self, path):
        os.makedirs(os.path.dirname(path) if os.path.dirname(path) else '.',
                    exist_ok=True)
        torch.save({
            'model_state_dict': self.state_dict(),
            'arch': 'SigilNetGraph',
        }, path)

    @classmethod
    def load(cls, path, device='cpu'):
        net = cls()
        ckpt = torch.load(path, map_location=device, weights_only=True)
        state = ckpt['model_state_dict']
        # Resize turn-feature-keyed projections to match an older
        # checkpoint's TURN_FEATURE_DIM if it differs from the current
        # constant — same approach as SigilNet.load.
        if 'turn_proj.weight' in state:
            d = state['turn_proj.weight'].shape[1]
            if d != net.turn_proj.in_features:
                net.turn_proj = nn.Linear(d, POLICY_HIDDEN_DIM)
        if 'blunder_turn_proj.weight' in state:
            d = state['blunder_turn_proj.weight'].shape[1]
            if d != net.blunder_turn_proj.in_features:
                from ai.sigil_net import BLUNDER_HIDDEN_DIM
                net.blunder_turn_proj = nn.Linear(d, BLUNDER_HIDDEN_DIM)
        net.load_state_dict(state, strict=False)
        net.eval()
        return net


class _DenseResBlock(nn.Module):
    """Pre-activation residual block, mirrors ai/sigil_net.py:ResBlock."""

    def __init__(self, dim):
        super().__init__()
        self.fc1 = nn.Linear(dim, dim)
        self.ln1 = nn.LayerNorm(dim)
        self.fc2 = nn.Linear(dim, dim)
        self.ln2 = nn.LayerNorm(dim)

    def forward(self, x):
        out = self.ln1(self.fc1(x))
        out = F.relu(out)
        out = self.ln2(self.fc2(out))
        return F.relu(out + x)
